### PR TITLE
MULE-9479: Add support for Datasense in Sources mode

### PR DESCRIPTION
**Pull request by estebanwasinger**

### CODE REVIEW

File: core/src/main/java/org/mule/api/metadata/MuleMetadataManager.java [hunk 1]
```diff
@@ -41,9 +42,9 @@
 public class MuleMetadataManager implements MetadataManager, Initialisable
 {
 
-    private static final String EXCEPTION_RESOLVING_OPERATION_METADATA = "An exception occurred while resolving Operation %s metadata";
-    private static final String PROCESSOR_NOT_METADATA_AWARE = "Operation is not MetadataAware, no information available";
-    private static final String EXCEPTION_RESOLVING_METADATA_KEYS = "An exception occurred while resolving Operation MetadataKeys";
+    private static final String EXCEPTION_RESOLVING_COMPONENT_METADATA = "An exception occurred while resolving Component %s metadata";
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/api/metadata/MuleMetadataManager.java]: rephrase to `... exception ocurred while resolving metadata for component '%s'`

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/metadata/SourceMetadataTestCase.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        final ComponentMetadataDescriptor componentMetadata = getComponentDynamicMetadata(personKey);
+        assertThat(componentMetadata.getContentMetadata().isPresent(), is(false));
+    }
+
+    @Test
+    public void getSourceStaticOutputMetadata() throws IOException
+    {
+
+        final ComponentMetadataDescriptor componentMetadata = getComponentStaticMetadata();
+        assertExpectedOutput(componentMetadata.getOutputMetadata(), new TypeToken<Map<String, Object>>()
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/metadata/SourceMetadataTestCase.java]: do not depend on gson just to do tests like this

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/metadata/SourceMetadataTestCase.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        componentId = new SourceId(SOURCE_METADATA);
+    }
+
+    @Test
+    public void getSourceMetadataKeys()
+    {
+        final MetadataResult<List<MetadataKey>> metadataKeysResult = metadataManager.getMetadataKeys(componentId);
+        assertThat(metadataKeysResult.isSuccess(), is(true));
+        final List<MetadataKey> metadataKeys = metadataKeysResult.get();
+        assertThat(metadataKeys.size(), is(3));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/metadata/SourceMetadataTestCase.java]: shouldn't you also tests that the keys are correct?

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java [hunk 4]
```diff
@@ -297,14 +300,22 @@ private void declareMessageSource(HasSourceDeclarer declarer, Class<? extends So
...
                 .withAttributesOfType(typeLoader.load(sourceGenerics.get(1)))
                 .withExceptionEnricherFactory(getExceptionEnricherFactory(sourceType))
-                .withModelProperty(new ImplementingTypeModelProperty(sourceType));
+                .withModelProperty(new ImplementingTypeModelProperty(sourceType))
+                .withMetadataResolverFactory(getMetadataResolverFactory(extensionType, sourceType));
+
+        declareSingleParameters(getParameterFields(sourceType), source, (ModelPropertyContributor) MuleExtensionAnnotationParser::parseMetadataAnnotations);
+        declareParameterGroups(sourceType, source);
 
-        declareAnnotatedParameters(sourceType, source);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java]: why?

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java [hunk 5]
```diff
@@ -297,14 +300,22 @@ private void declareMessageSource(HasSourceDeclarer declarer, Class<? extends So
...
     }
 
     private void declareAnnotatedParameters(Class<?> annotatedType, ParameterizedDeclarer parameterDeclarer)
     {
         declareSingleParameters(getParameterFields(annotatedType), parameterDeclarer);
+        declareParameterGroups(annotatedType, parameterDeclarer);
+    }
+
+    private void declareParameterGroups(Class<?> annotatedType, ParameterizedDeclarer parameterDeclarer)
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java]: why?

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/metadata/SourceMetadataTestCase.java (hunk 2 above)
Comment (estebanwasinger) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/metadata/SourceMetadataTestCase.java]: is not Gson, is Guava

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/ExtensionComponent.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    @Override
+    public void setMuleContext(MuleContext context)
+    {
+        this.muleContext = context;
+    }
+
+    /**
+     * 2
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/ExtensionComponent.java]: 2 what?

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/ExtensionComponent.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                                                           ? extensionManager.getConfigurationProvider(extensionModel)
+                                                           : extensionManager.getConfigurationProvider(configurationProviderName);
+        return provider;
+    }
+
+    /**
+     * Validates if the current component is valid for the set configuration.
+     * In case that the validation fails, the method will throw a {@link IllegalComponentException}
+     */
+    protected void validateComponentConfiguration()
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/ExtensionComponent.java]: This code is overly complicated in comparison to the one you're replacing. It will also be very hard for it to provide a good error message.

I would rather have each subclass implementing this on its own way than trying to have this generic logic.

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/processor/IllegalComponentException.java [hunk 8]
```diff
@@ -7,20 +7,20 @@
...
 
 /**
- * A {@link RuntimeException} which marks that a selected operation
+ * A {@link RuntimeException} which marks that a selected component
  * is not valid or misconfigured
  *
  * @since 4.0
  */
-public class IllegalOperationException extends RuntimeException
+public class IllegalComponentException extends RuntimeException
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/processor/IllegalComponentException.java]: I would rather have IllegalOperaitonException extend IllegalComponentException, but it's still good to have qualified errors for invalid source and invalid operations. That makes error handling and testing easier.

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/processor/OperationMessageProcessor.java [hunk 9]
```diff
@@ -162,36 +111,13 @@ private OperationContextAdapter createOperationContext(ConfigurationInstance<Obj
     @Override
     public void initialise() throws InitialisationException
     {
+        validateComponentConfiguration();
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/processor/OperationMessageProcessor.java]: this code is not equivalent to the one being replaced. It is possible for the correct configuration provider to not be availabel at initialisation time which is why the getConfiguration() method you're removing has an orElseGet() call. Revert the changes associated to that piece. Leave each component type to resolve that on its own way

Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/processor/OperationMessageProcessor.java]: disregard this comment

Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/processor/OperationMessageProcessor.java]: but do leave each component type to validate itself on its own way

File: modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/metadata/SourceMetadataTestCase.java (hunk 2 above)
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/module/extension/internal/metadata/SourceMetadataTestCase.java]: disregard then

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java (hunk 4 above)
Comment (estebanwasinger) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java]: I opened the declaration of source parameters to be more flexible and be able to contribute with ModelProperties at declaration time

File: modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java (hunk 5 above)
Comment (estebanwasinger) [inline modules/extensions-support/src/main/java/org/mule/module/extension/internal/introspection/describer/AnnotationsBasedDescriber.java]: I group the logic to be more declarative and re use it in the Source parameters declaration
